Add checklist item reordering via dnd-kit

clavidk opened the pull request:
fixes #99 

#### Summary
Adds drag-and-drop reordering for checklist items within a note on the board page, with optimistic UI updates and persistence. The feature code has been refactored for clarity and reusability.

#### What changed
- Components
  - `components/checklist/ChecklistList.tsx`: Owns DnD context, sensors, vertical-only modifiers, and list orchestration.
  - `components/checklist/SortableChecklistItem.tsx`: Item UI + drag wiring; supports edit, split on Enter, delete, and toggle.
- Hook
  - `hooks/useDebouncedChecklistEdit.ts`: Centralizes debounced edit behavior; exposes `debouncedEdit`, `cancelPendingEdit`, and `clearAll`.
- Utilities
  - `lib/checklist.ts`: Shared `ChecklistItem` type and helpers for future reuse.
- Page integration
  - `app/boards/[id]/page.tsx`: Replaced inline DnD and ad‑hoc timeout map with the new components/hook. All existing board/notes logic, filters, and API calls remain unchanged.
- Dependencies
  - `@dnd-kit/core`, `@dnd-kit/sortable`, `@dnd-kit/utilities`, `@dnd-kit/modifiers` already added in this branch.

#### UX
- Drag items within a note to reorder.
- Order updates instantly (optimistic), then persists via `PUT /api/boards/{boardId}/notes/{noteId}` with updated `checklistItems` order and `done`.
- “Done” state recalculates based on all items being checked.

Before:
<img width="434" height="219" alt="image" src="https://github.com/user-attachments/assets/d55dd39e-d213-4adf-86d4-b458b771c293" />

After:
![fixed](https://github.com/user-attachments/assets/f63f733b-d3a1-47bc-8285-240d5f73b77a)

#### Implementation details
- DnD
  - Context and sensors encapsulated in `ChecklistList` for easier tuning.
  - Modifiers: `restrictToVerticalAxis`, `restrictToParentElement`.
  - Activation constraint: `PointerSensor` distance set to 4 (balance between snappy start and avoiding accidental drags).
  - Keyboard sensor included for future accessibility improvements.
- Editing
  - Debounced edits via `useDebouncedChecklistEdit`; pending timers are cancelled on blur and cleared when exiting edit mode.
- Performance
  - Avoids heavy re-renders by localizing dnd-kit hooks to list/item components.
  - Uses `transform-gpu` and removes transform transitions while dragging to prevent “trailing.”

#### Files touched
- `components/checklist/ChecklistList.tsx` (new)
- `components/checklist/SortableChecklistItem.tsx` (new)
- `hooks/useDebouncedChecklistEdit.ts` (new)
- `lib/checklist.ts` (new)
- `app/boards/[id]/page.tsx` (updated to consume components/hook)
- `package.json`, `package-lock.json` (dependencies added earlier in the branch)

#### Testing
- Manual
  - Create a note with multiple checklist items and reorder; refresh to verify persistence.
  - Toggle items and confirm “done” recalculates correctly.
  - Verify author/admin permissions for editing and deletion.
- Automated
  - Existing unit tests pass; no backend changes required.

#### Limitations / follow-ups
- Reordering is within a single note; cross-note drag is out-of-scope.
- Consider adding e2e coverage for drag reorder behavior.
- Future: expose a read-only mode using the same components for public boards; add keyboard sorting.

#### Backward compatibility
- No API/migration changes.
- Non-DnD board logic remains intact.

slavingia commented:
Could we remove the drag handles and just use click-and-drag?

Like: https://master--5fc05e08a4a65d0021ae0bf2.chromatic.com/?path=/story/presets-sortable-vertical--without-drag-overlay

clavidk commented:
> Could we remove the drag handles and just use click-and-drag?

Done:
![2025-08-07 at 11 58 16](https://github.com/user-attachments/assets/1eff9ae7-1c0b-4e54-b10e-8052c9cf91f0)

slavingia commented:
I'll test tmmrw but that recording seems a little laggy (I had similar issues locally)

clavidk commented:
> seems a little laggy (I had similar issues locally)

Fixed it, very smooth and zippy now (any lag in the GIF is probably b/c it's a GIF):
![fixed](https://github.com/user-attachments/assets/f63f733b-d3a1-47bc-8285-240d5f73b77a)


Also fixed an issue where dragging could let you drag horizontally triggering an ugly horizontal scroll bar and an expanding canvas/overflow. This was before:
![horizontal-scroll-bug](https://github.com/user-attachments/assets/55b9836b-4f33-4131-92ff-2176277c734a)

slavingia commented:
Please merge `main` and resolve merge conflicts. We've component-ized the crazy big board page which should help a lot with moving fast and testing things more easily.

slavingia commented on app/api/boards/[id]/notes/[noteId]/route.ts:
```diff
@@ -2,7 +2,7 @@ import { NextRequest, NextResponse } from "next/server"
 import { db } from "@/lib/db"
 import { updateSlackMessage, formatNoteForSlack, sendSlackMessage, sendTodoNotification, hasValidContent, shouldSendNotification } from "@/lib/slack"
-import type { ChecklistItem } from "@/components/checklist-item"
+import type { ChecklistItem } from "@/lib/types/checklist-item"
```
Why this change? Introduces a new file that IMO doesn't add much.

slavingia commented on app/boards/[id]/page.tsx:
```diff
@@ -80,6 +80,8 @@ export default function BoardPage({
   const router = useRouter();
   const searchParams = useSearchParams();
 
+  // DnD sensors reused for all note lists
```
?

slavingia commented on app/boards/[id]/page.tsx:
```diff
@@ -480,6 +482,7 @@ export default function BoardPage({
     addingChecklistItem,
   ]);
 
+  // Removed debounce cleanup effect; editing is scoped to Note
```
Maybe not ready for review. Will wait for a ping

clavidk commented:
@slavingia looks like there's already another cleaner PR open for this so I'll just let you focus on that one and close this one out. Was good practice for me, will try on another issue next time!
